qiime/workflow/preprocess.py

```diff
@@ -148,22 +164,56 @@ def create_commands_slf(all_files, demultiplexing_method, output_dir,
             barcode_files.append(all_files[curr_fp][0])
             mapping_files.append(all_files[curr_fp][1])
 
+    # gather all the arguments into a dictionary to format them into a command
+    params_dict = {'leading_text': _clean_leading_text(leading_text),
+                   'trailing_text': trailing_text, 'output_dir': output_dir,
+                   'params': params,
+                   'read_files': _make_file(read_files,
+                                            path_builder('seqs.txt'))}
+    cmd = ("{leading_text}split_libraries_fastq.py {params} "
+           "-i {read_files} -o {output_dir} --read_arguments_from_file ")
+
     if demultiplexing_method == 'sampleid_by_file':
-        command =\
-            "%ssplit_libraries_fastq.py %s -i %s --sample_ids %s -o %s %s --barcode_type 'not-barcoded'" %\
-            (_clean_leading_text(leading_text), params, ",".join(read_files), ",".join(sample_ids),
-            output_dir, trailing_text)
+        cmd += ("--sample_ids {sample_ids} --barcode_type 'not-barcoded' ")
+        params_dict['sample_ids'] = _make_file(sample_ids,
+                                               path_builder('sample_ids.txt'))
     else:
-        command =\
-            "%ssplit_libraries_fastq.py %s -i %s --barcode_read_fps %s --mapping_fps %s -o %s %s" %\
-            (_clean_leading_text(leading_text), params, ",".join(read_files),
-            ",".join(barcode_files), ",".join(mapping_files),
-            output_dir, trailing_text)
+        cmd += ("--barcode_read_fps {barcode_files} "
+                "--mapping_fps {mapping_files} ")
+        params_dict['barcode_files'] = _make_file(barcode_files,
+                                                  path_builder('barcodes.txt'))
+        params_dict['mapping_files'] = _make_file(mapping_files,
+                                                  path_builder('maps.txt'))
 
-    commands.append([('split_libraries_fastq.py', command)])
+    # regardless of the method, trailing_text should go last
+    cmd += '{trailing_text}'
+    commands.append([('split_libraries_fastq.py', cmd.format(**params_dict))])
 
     return commands
 
+
+def _make_file(elements, fp):
+    """Create a file with one element per line
+
+    This function is just a helper for create_commands_slf.
+
+    Parameters
+    ----------
+    elements : list
```

Comment:
Yes, though in the context of create_commands_slf, it should only take
ordered iterables, thus my decision to make it `list`. Otherwise
sequences and sample names could get out of sync :scream: :scream:
What about changing this to list, tuple?

On (Aug-28-15|16:26), adamrp wrote:
> >  
> >      return commands
> >  
> > +
> > +def _make_file(elements, fp):
> > +    """Create a file with one element per line
> > +
> > +    This function is just a helper for create_commands_slf.
> > +
> > +    Parameters
> > +    ----------
> > +    elements : list
> 
> more generally, could the type be any `iterable`?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/biocore/qiime/pull/2077/files#r38253935